docs/langref/relay_type.rst

```diff
@@ -0,0 +1,222 @@
+===================
+Relay's Type System
+===================
+
+We have briefly introduced types while detailing the the expression language
+of Relay, but have not yet described the type system. Relay is
+a statically typed and type-inferred language, allowing programs to
+be typed with a minimal requirement of explicit type annotations.
+
+Static types are useful when performing compiler optimizations because they
+communicate properties about the data a program manipulates, such as runtime
+shape, data layout, and storage, without needing to run the program.
+
+Relay's type system features a form of dependent typing for shapes. Treating tensor
+shapes as types allows Relay to perform more powerful reasoning at compile time;
+in particular, Relay can statically reason about operations whose output shapes
+vary based on the input shapes in complex ways. Casting shape inference as a type
+inference problem allows Relay to infer the shapes of all tensors at compile time,
+including in programs that use branching and function calls.
+
+Statically reasoning about shapes in this manner allows
+Relay to be ahead-of-time compiled and provides much more information about
+tensors for optimizations further in the compilation pipeline. Such optimizations
+can be implemented as passes, which are Relay-to-Relay AST transformations, and
+may use the inferred types (e.g., shape information) for making decisions about
+program transformations. For instance, `src/relay/pass/fuse_ops.cc` gives
+an implementation of a pass that uses inferred tensor shapes to replace invocations
+of operators in a Relay program with fused operator implementations.
+
+Reasoning about tensor types in Relay is encoded using type relations, which means
+that the bulk of type checking in Relay is constraint solving (ensuring that all
+type relations are satisfied at call sites). Type relations offer a flexible and
+relatively simple way of making the power of dependent typing available in Relay
+without greatly increasing the complexity of Relay's type system.
+
+Types
+=====
+
+Below we detail the language of types in Relay and how they are assigned to Relay expressions.
+
+Type
+~~~~
+
+The base type for all Relay types. All Relay types are sub-classes of this base type.
+
+See :py:class:`~tvm.relay.ty.Type` for its definition and documentation.
+
+Tensor Type
+~~~~~~~~~~~
+
+A concrete tensor type in Relay.
+
+Tensors are typed according to data type and shape. At present, these use TVM's
+data types and shapes, but in the future, Relay may include a separate AST for
+shapes. In particular, data types include `bool`, `float32`, `int8` and various
+other bit widths. Shapes are given as tuples of dimensions (TVM `IndexExpr`),
+such as `(5, 5)`; scalars are also given tuple types and have a shape of `()`.
+
+Note, though, that TVM shapes can also include variables and arithmetic expressions
```

Comment:
this could benefit from some examples. we should probably have an example of these non-trivial types and also an example of where shapes must be concrete, especially how it relates to polymorphism.

Reply:
I'll see if there's a link to pre-existing documentation of that feature, as it's not new to Relay

Reply:
I couldn't find a link to a convenient description of NNVM's existing shape inference, but I hope someone knows of one, because I am not especially familiar with it and would rather not retread old ground when I imagine there are many people who are much more knowledgeable about that mechanism.